pytorch_lightning/metrics/classification/utils.py

```diff
@@ -0,0 +1,390 @@
+            " and preds (N, C, ...) or (N, ..., C)"
+        )
+
+    if preds.shape != target.shape and is_multiclass is False and extra_dim_size != 2:
+        raise ValueError(
+            "You have set is_multiclass=False, but have more than 2 classes in your data,"
+            " based on the C dimension of preds."
+        )
+
+    # Check that num_classes is consistent
+    if not num_classes:
+        if preds.shape != target.shape and target.max() >= extra_dim_size:
+            raise ValueError("The highest label in targets should be smaller than the size of C dimension")
+    else:
+        if case == "binary":
+            if num_classes > 2:
+                raise ValueError("Your data is binary, but num_classes is larger than 2.")
+            elif num_classes == 2 and not is_multiclass:
+                raise ValueError(
+                    "Your data is binary and num_classes=2, but is_multiclass is not True."
+                    "Set it to True if you want to transform binary data to multi-class format."
+                )
+            elif num_classes == 1 and is_multiclass:
+                raise ValueError(
+                    "You have binary data and have set is_multiclass=True, but num_classes is 1."
+                    "Either leave is_multiclass unset or set it to 2 to transform binary data to multi-class format."
+                )
+        elif "multi-class" in case:
+            if num_classes == 1 and is_multiclass is not False:
+                raise ValueError(
+                    "You have set num_classes=1, but predictions are integers."
+                    "If you want to convert (multi-dimensional) multi-class data with 2 classes"
+                    "to binary/multi-label, set is_multiclass=False."
+                )
+            elif num_classes > 1:
+                if is_multiclass is False:
+                    if implied_classes != num_classes:
+                        raise ValueError(
+                            "You have set is_multiclass=False, but the implied number of classes "
+                            "(from shape of inputs) does not match num_classes. If you are trying to"
+                            "transform multi-dim multi-class data with 2 classes to multi-label, num_classes"
+                            "should be either None or the product of the size of extra dimensions (...)."
+                            "See Input Types in Metrics documentation."
+                        )
+                if num_classes <= target.max():
+                    raise ValueError("The highest label in targets should be smaller than num_classes")
+                if num_classes <= preds.max():
+                    raise ValueError("The highest label in preds should be smaller than num_classes")
+                if preds.shape != target.shape and num_classes != extra_dim_size:
+                    raise ValueError("The size of C dimension of preds does not match num_classes")
```

Comment:
just thinking to wrap this into another protected function for:
- shorten the func as it is quite long and all the if/else are uneasy to follow
- give the functional block a name so easier to understand the whole context

Reply:
Ok, I can create some sub-functions here for each case

Reply:
Doneio :)